### DISCUSSION TITLE
MagickImage Native Memory Management

### DISCUSSION THREAD

**Discussion by jkavanaghrdm** [dlemstra/Magick.NET#857] [Help]
We are using MagickImage Load, Resize and ToByteArray operations and are finding that Working Set memory on CentOS 7 does not get deallocated when those operations complete and the MagickImage object is disposed. Is that the expected behavior? Or does MagickImage maintain an internal native memory pool for it's own purposes? I think the fundamental question is, should native memory allocations be fully deallocated once the MagickImage object is disposed?

If there are any specific details missing to be able to answer this question then please let me know.

**Comment by dlemstra** [comment]
There is no internal native memory pool for ImageMagick, I suspect that the memory that you are seeing is managed memory instead. The garbage collector will probably keep some memory in it's pool.

**Reply by jkavanaghrdm** [reply]
Thanks for the response.

We do check the dotnet-counters command output to check managed and unmanaged memory consumption and see output like the below which indicates that managed heap is fine (GC Heap Size) but the Working Set is elevated (Working Set (MB)):

System.Runtime
--
% Time in GC since last GC (%)                                 1
Allocation Rate (B / 1 sec)                               65,344
CPU Usage (%)                                                    0
Exception Count (Count / 1 sec)                                0
**GC Heap Size (MB)                                               15**
Gen 0 GC Count (Count / 1 sec)                                 0
Gen 0 Size (B)                                                  48
Gen 1 GC Count (Count / 1 sec)                                 0
Gen 1 Size (B)                                              59,792
Gen 2 GC Count (Count / 1 sec)                                 0
Gen 2 Size (B)                                           3,667,000
LOH Size (B)                                             9,792,000
Monitor Lock Contention Count (Count / 1 sec)                  0
Number of Active Timers                                       23
Number of Assemblies Loaded                                   94
ThreadPool Completed Work Item Count (Count / 1 sec)           4
ThreadPool Queue Length                                        0
ThreadPool Thread Count                                       15
**Working Set (MB)                                               319**

We are finding that the Working Set does not return to the expected values after we finish our test. Any thoughts on why that is?

**Reply by dlemstra** [reply]
Your issues looks similar to this: https://github.com/dlemstra/Magick.NET/issues/707. Haven't heared back from them and they also thought this was something in the garbage collector.

**Reply by jkavanaghrdm** [reply]
I posted to issue #707 for feedback as well.
